[Validator] Add `Xml` constraint

sfmok opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 7.3
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Issues        | No
| License       | MIT

Add `Xml` constraint

OskarStark commented:
Did you check
* https://github.com/symfony/symfony/pull/53749

?

stof commented on src/Symfony/Component/Validator/Constraints/AbstractFormat.php:
```diff
@@ -0,0 +1,32 @@
+    public string $format;
+    public string $message = 'This value should be valid {{ format }}.';
+
+    public function __construct(?string $message = null, ?array $options = null, ?array $groups = null, mixed $payload = null)
```
either `$options` must be removed entirely and the constructor should have the HasNamedParameter attribute (supporting only the modern way of defining constraints), or `$options` must be the first argument of the constructor (to be compatible with the legacy way of defining them).

Right now, this constructor signature is incompatible with using this constraint in the YAML or XML file loaders.

stof replied:
Note that the `Json` constraint **must** keep support for the legacy way of defining constraint as it is already released with that support.

sfmok replied:
Thanks @stof, I have removed the `format` abstraction and revert `json` constraint changes

stof commented on src/Symfony/Component/Validator/Constraints/AbstractFormat.php:
```diff
@@ -0,0 +1,32 @@
+abstract class AbstractFormat extends Constraint
+{
+    public string $format;
+    public string $message = 'This value should be valid {{ format }}.';
```
should we really using a `format` placeholder here ? This might look weird when translating error messages as the error message won't be translated. I suggest using separate default messages in each constraint instead.

stof replied:
My suggestion would be to remove this AbstractFormat class entirely. Having a based class for those constraint is not really worth it IMO.

alexandre-daubois replied:
To support this point, it's very subjective, but I find it very hard to see JWT as a format on a par with Yaml and Xml.

sfmok replied:
Done

stof commented on src/Symfony/Component/Validator/Constraints/Json.php:
```diff
@@ -11,32 +11,28 @@
         self::INVALID_JSON_ERROR => 'INVALID_JSON_ERROR',
     ];
 
-    public string $message = 'This value should be valid JSON.';
```
changing the default message of this constraint will break any project relying on translating this message.

stof commented on src/Symfony/Component/Validator/Constraints/JwtValidator.php:
```diff
@@ -0,0 +1,41 @@
+        $decodedHeader = base64_decode(strtr($header, '-_', '+/'));
+        $decodedPayload = base64_decode(strtr($payload, '-_', '+/'));
+
+        return json_validate($decodedHeader) && json_validate($decodedPayload);
```
I don't think this is enough to validate that this is a valid JWT. Having a core validator doing a *partial* validation is worse than not having it IMO.

alexandre-daubois replied:
Agree. Some may rely on a custom constraint to invoke `lcobucci/jwt` (for example) to ensure everything's in place, including verifying the token signature.

I understand that being able to validate a JWT with a constraint is nice, but I would expect a constraint like this to check signature/issuer/etc. (which would really complexify the logic in the validator), otherwise you would need to write a second checker anyway to do some asserts on info included in the token. This would then void the usefulness of such constraint.

sfmok replied:
Thank you all for your valuable feedback. I will remove the JWT constraint for now and might push a separate PR for it later, as it requires extensive work to verify the signature.

alexandre-daubois commented on src/Symfony/Component/Validator/Tests/Constraints/YamlValidatorTest.php:
```diff
@@ -0,0 +1,117 @@
+    public static function getValidYamlValues(): array
+    {
+        $yaml1 = <<<YAML
+string: "value with special characters: !@#$%^&*()"
+multiline: |
+  This is a
+  multiline
+  string.
+YAML;
+        $yaml2 = <<<YAML
+key: value
+#address
+address:
+  items:
+    - item1
+    - item2
+    - item3
+  address:
+    country: Tunisia
+    city: Kebili
+    zip_code: 4230
+YAML;
```
If I remember correctly, you're allowed in newer PHP versions to indent your Heredoc with the rest of your code

alexandre-daubois commented on src/Symfony/Component/Validator/Tests/Constraints/YamlValidatorTest.php:
```diff
@@ -0,0 +1,117 @@
+
+    public static function getInvalidYamlValues(): array
+    {
+        $yaml = <<<YAML
+key: value
+#address
+address:
+  items:
+    - item1
+    item2
+    - item3
+  address:
+    country: Tunisia
+    city:Kebili
+    zip code: 4230
+YAML;
```
Same here

alexandre-daubois commented on src/Symfony/Component/Validator/Constraints/AbstractFormatValidator.php:
```diff
@@ -0,0 +1,57 @@
+            ->addViolation();
+    }
+
+    /**
+     * Validate format value.
+     */
```
Not sure this comment helps a lot understand what the method's doing 😄 Should it be removed?

OskarStark commented:
`Yaml` constraint was merged

OskarStark commented on src/Symfony/Component/Validator/CHANGELOG.md:
```diff
@@ -3,7 +3,7 @@ CHANGELOG
 7.2
 ---
-
+ * Add the `Yaml`, `Xml` and `Jwt` constraints
```
```suggestion
 * Add the `Xml` and `Jwt` constraints
```

sfmok commented:
I've made the following changes:

- Removed format abstraction
- Removed the `Yaml` constraint from this PR since it has already been merged
- Reverted the `Json` constraint changes to prevent breaking the translation
- Remove the `Jwt` constraint for now, as verifying the signature requires extensive work. I will likely push a separate PR for it later.
- Adjusted the `Xml` constraint to support only the modern method of defining constraints

OskarStark reviewed (approved):
Needs a rebase

alexandre-daubois commented on src/Symfony/Component/Validator/CHANGELOG.md:
```diff
@@ -3,7 +3,7 @@ CHANGELOG
 7.2
 ---
-
+ * Add the `Xml` constraint for validating XML content
```
New lines should be appended, not prepended

alexandre-daubois commented on src/Symfony/Component/Validator/Constraints/Xml.php:
```diff
@@ -0,0 +1,39 @@
+use Symfony\Component\Validator\Constraint;
+
+/**
+ * Validates that a value is a valid xml format.
```
```suggestion
 * Validates that a value is a valid XML string.
```

alexandre-daubois commented on src/Symfony/Component/Validator/Tests/Constraints/XmlValidatorTest.php:
```diff
@@ -0,0 +1,89 @@
+    public static function getValidXmlValues(): array
+    {
+        $xml = <<<XML
+<?xml version="1.0" encoding="UTF-8"?>
+<srv:container xmlns="http://symfony.com/schema/dic/security"
+    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xmlns:srv="http://symfony.com/schema/dic/services"
+>
+    <config>
+        <provider name="default">
+            <memory>
+                <user identifier="foo" password="foo" />
+            </memory>
+        </provider>
+        <firewall name="simple" pattern="/" security="false" />
+    </config>
+</srv:container>
+XML;
```
This can be indented with the rest 🙂

alexandre-daubois reviewed (commented):
Just a few nitpicks 🙂 Nice PR!

nicolas-grekas commented on src/Symfony/Component/Validator/CHANGELOG.md:
```diff
@@ -3,11 +3,11 @@ CHANGELOG
 
 7.2
 ---
-
```
should be kept

sfmok replied:
Done

nicolas-grekas reviewed (commented):
Asserting that a file is XML is pretty weak as a check: one always expects some structure in the XML. It'd be natural to me to allow passing an XSD, isn't it? Up to improve the PR in this direction?

OskarStark commented on src/Symfony/Component/Validator/Tests/Constraints/XmlValidatorTest.php:
```diff
@@ -0,0 +1,89 @@
+    /**
+     * @dataProvider getInvalidXmlValues
+     */
+    public function testInValidXmlValue($value)
```
```suggestion
    public function testInvalidXmlValue(string $value)
```

OskarStark commented on src/Symfony/Component/Validator/Tests/Constraints/XmlValidatorTest.php:
```diff
@@ -0,0 +1,89 @@
+     */
+    public function testInValidXmlValue($value)
+    {
+        $constraint = new Xml(message: 'myMessage');
+        $this->validator->validate($value, $constraint);
```
```suggestion
        $this->validator->validate($value, new Xml(message: 'myMessage'));
```

sfmok commented:
> Asserting that a file is XML is pretty weak as a check: one always expectes some structure in the XML. It'd be natural to me to allow passing an XSD, isn't it? Up to improve the PR in this direction?

@nicolas-grekas Thank you for your feedback. The constraint is intended to assert a string value rather than a file.
I'm not sure about allowing XSD checks, but I will consider supporting it if we receive more positive feedback either before or after merging the PR

94noni reviewed (approved):
Nice

sfmok commented:
Any updates?

fabpot commented:
> > Asserting that a file is XML is pretty weak as a check: one always expectes some structure in the XML. It'd be natural to me to allow passing an XSD, isn't it? Up to improve the PR in this direction?
> 
> @nicolas-grekas Thank you for your feedback. The constraint is intended to assert a string value rather than a file. I'm not sure about allowing XSD checks, but I will consider supporting it if we receive more positive feedback either before or after merging the PR

I also think that supporting validating the XML content against a schema would make this feature more useful.

nicolas-grekas reviewed (changes requested):
Status: needs work

Still up to continue the PR?

sfmok commented:
@nicolas-grekas Yes, I'll work on it when I have some free time.

sfmok commented:
@nicolas-grekas & @fabpot: Thanks again for the feedback. XSD schema validation is now supported.

alexandre-daubois commented on src/Symfony/Component/Validator/Constraints/Xml.php:
```diff
@@ -0,0 +1,47 @@
+    ) {
+        parent::__construct(null, $groups, $payload);
+
+        if (null !== $this->schemaPath && !file_exists($this->schemaPath)) {
```
Let's use `is_readable()` to ensure that the file exists and can actually be read:

```suggestion
        if (null !== $this->schemaPath && !is_readable($this->schemaPath)) {
```

GromNaN replied:
In this case, you also need to change the error message: "file does not exist or cannot be read".

sfmok replied:
Thanks for your valuable feedback. Resolved

sfmok committed:
[Validator] Add xml validator

nicolas-grekas reviewed (approved):
I rebased on 8.1

GromNaN commented on src/Symfony/Component/Validator/Constraints/XmlValidator.php:
```diff
@@ -0,0 +1,75 @@
+            if ($constraint->schemaPath) {
+                $dom = \dom_import_simplexml($doc)->ownerDocument;
+
+                if (!$dom->schemaValidate($constraint->schemaPath, $constraint->schemaFlags)) {
```
If the XSD file itself is invalid, this reports as contraint violation "This value is not conform to the expected XSD schema.". This should be a regular exception, that's an issue with the constraint configuration, not the value.

nicolas-grekas replied:
addressed!

nicolas-grekas commented:
Thank you @sfmok.
